File: src/core/genome.ts
/**
 * El genoma: una semilla de 64 bits que ES la criatura.
 *
 * No hay tabla de mascotas ni lista de sprites. Cuerpo, color, carácter y
 * rarezas se derivan todos de este número. Mismo número → misma criatura,
 * siempre. Compartís el seed y la otra persona ve exactamente lo mismo.
 */

export const GENOME_BITS = 64n;
const MASK64 = (1n << GENOME_BITS) - 1n;

/** Extrae `length` bits del genoma a partir de `offset`. */
function bits(seed: bigint, offset: number, length: number): number {
  const mask = (1n << BigInt(length)) - 1n;
  return Number((seed >> BigInt(offset)) & mask);
}

export interface StatBias {
  vigor: number;
  animo: number;
  ingenio: number;
  vinculo: number;
}

export interface Genes {
  /** Linaje base (0-15). Define el "de qué familia es". */
  lineage: number;
  /** Arquetipo de silueta (0-15). */
  bodyShape: number;
  /** Estilo de ojos (0-15). */
  eyes: number;
  /** Estilo de boca (0-15). */
  mouth: number;
  /** Bitfield de apéndices: 1=orejas 2=cuernos 4=alas 8=cola. */
  appendages: number;
  /** Patrón sobre el cuerpo (0-15). */
  pattern: number;
  /** Tono base 0-255 → 0-360°. */
  hue: number;
  /** Modo de paleta (0-7). */
  paletteMode: number;
  /** Carácter (0-7). Afecta los eventos de la simulación. */
  temperament: number;
  /** Velocidad a la que gasta energía (0-7). */
  metabolism: number;
  /** Afinidad elemental (0-7). Sesga la rama evolutiva. */
  affinity: number;
  /** Proporciones de cabeza y cuerpo (0-15). */
  proportion: number;
  /** Sesgo de estadísticas, 2 bits cada una. */
  statBias: StatBias;
  /** Bits de mutación / reserva (0-255). */
  mutation: number;
}

/**
 * Mapa de bits del genoma. El orden importa: cambiarlo invalida todos los
 * seeds existentes, así que se versiona junto con el formato de guardado.
 */
export const GENOME_LAYOUT = [
  { key: "lineage", offset: 0, bits: 4 },
  { key: "bodyShape", offset: 4, bits: 4 },
  { key: "eyes", offset: 8, bits: 4 },
  { key: "mouth", offset: 12, bits: 4 },
  { key: "appendages", offset: 16, bits: 4 },
  { key: "pattern", offset: 20, bits: 4 },
  { key: "hue", offset: 24, bits: 8 },
  { key: "paletteMode", offset: 32, bits: 3 },
  { key: "temperament", offset: 35, bits: 3 },
  { key: "metabolism", offset: 38, bits: 3 },
  { key: "affinity", offset: 41, bits: 3 },
  { key: "proportion", offset: 44, bits: 4 },
  { key: "statBias", offset: 48, bits: 8 },
  { key: "mutation", offset: 56, bits: 8 },
] as const;

export function decodeGenome(seed: bigint): Genes {
  const normalized = seed & MASK64;
  const statRaw = bits(normalized, 48, 8);

  return {
    lineage: bits(normalized, 0, 4),
    bodyShape: bits(normalized, 4, 4),
    eyes: bits(normalized, 8, 4),
    mouth: bits(normalized, 12, 4),
    appendages: bits(normalized, 16, 4),
    pattern: bits(normalized, 20, 4),
    hue: bits(normalized, 24, 8),
    paletteMode: bits(normalized, 32, 3),
    temperament: bits(normalized, 35, 3),
    metabolism: bits(normalized, 38, 3),
    affinity: bits(normalized, 41, 3),
    proportion: bits(normalized, 44, 4),
    statBias: {
      vigor: statRaw & 0b11,
      animo: (statRaw >> 2) & 0b11,
      ingenio: (statRaw >> 4) & 0b11,
      vinculo: (statRaw >> 6) & 0b11,
    },
    mutation: bits(normalized, 56, 8),
  };
}

// ---------------------------------------------------------------------------
// Nombres. Nada de "perrito" o "gatito": estas criaturas no son animales.
// ---------------------------------------------------------------------------

export const LINEAGES = [
  "Nébula",
  "Fungo",
  "Cristal",
  "Limo",
  "Pluma",
  "Escama",
  "Musgo",
  "Brasa",
  "Vapor",
  "Óxido",
  "Coral",
  "Cirro",
  "Duna",
  "Eco",
  "Prisma",
  "Raíz",
] as const;

export const TEMPERAMENTS = [
  "Plácido",
  "Curioso",
  "Arisco",
  "Leal",
  "Errante",
  "Voraz",
  "Tímido",
  "Feroz",
] as const;

export const AFFINITIES = [
  "Brasa",
  "Marea",
  "Raíz",
  "Chispa",
  "Escarcha",
  "Polvo",
  "Eco",
  "Vacío",
] as const;

export const METABOLISMS = [
  "Aletargado",
  "Lento",
  "Sereno",
  "Regular",
  "Activo",
  "Inquieto",
  "Ávido",
  "Frenético",
] as const;

/** Devuelve un elemento de una tabla de nombres sin salirse del rango. */
function nameFrom(table: readonly string[], index: number): string {
  return table[index % table.length] ?? table[0] ?? "?";
}

export function lineageName(genes: Genes): string {
  return nameFrom(LINEAGES, genes.lineage);
}

export function temperamentName(genes: Genes): string {
  return nameFrom(TEMPERAMENTS, genes.temperament);
}

export function affinityName(genes: Genes): string {
  return nameFrom(AFFINITIES, genes.affinity);
}

export function metabolismName(genes: Genes): string {
  return nameFrom(METABOLISMS, genes.metabolism);
}

// ---------------------------------------------------------------------------
// Serialización del seed
// ---------------------------------------------------------------------------

/** Formatea el seed como hex agrupado: `A3F0-91C4-77BE-2D08`. */
export function formatSeed(seed: bigint): string {
  const hex = (seed & MASK64).toString(16).toUpperCase().padStart(16, "0");
  return (hex.match(/.{4}/g) ?? [hex]).join("-");
}

/**
 * Interpreta la entrada del usuario como seed.
 *
 * Acepta hex con o sin guiones y con o sin `0x`, o un decimal. Cualquier otra
 * cosa se hashea a un seed válido, así que escribir tu nombre también funciona
 * y siempre da la misma criatura.
 */
export function parseSeed(input: string): bigint {
  const trimmed = input.trim();
  if (trimmed === "") {
    throw new Error("El seed no puede estar vacío");
  }

  const compact = trimmed.replace(/[\s-]/g, "");

  if (/^0x[0-9a-f]+$/i.test(compact)) {
    return BigInt(compact) & MASK64;
  }
  if (/^[0-9a-f]{1,16}$/i.test(compact) && /[a-f]/i.test(compact)) {
    return BigInt(`0x${compact}`) & MASK64;
  }
  if (/^\d+$/.test(compact)) {
    return BigInt(compact) & MASK64;
  }
  if (/^[0-9a-f]{1,16}$/i.test(compact)) {
    return BigInt(`0x${compact}`) & MASK64;
  }

  return hashString(trimmed);
}

/** Convierte texto arbitrario en un seed de 64 bits (FNV-1a de 64 bits). */
export function hashString(text: string): bigint {
  let hash = 0xcbf29ce484222325n;
  const prime = 0x100000001b3n;
  for (let i = 0; i < text.length; i++) {
    hash = ((hash ^ BigInt(text.charCodeAt(i))) * prime) & MASK64;
  }
  return hash;
}

/** Genera un seed nuevo con entropía criptográfica. */
export function randomSeed(): bigint {
  const bytes = new Uint8Array(8);
  globalThis.crypto.getRandomValues(bytes);
  let seed = 0n;
  for (const byte of bytes) {
    seed = (seed << 8n) | BigInt(byte);
  }
  return seed & MASK64;
}

/** Normaliza cualquier bigint al rango de 64 bits sin signo. */
export function normalizeSeed(seed: bigint): bigint {
  return ((seed % (MASK64 + 1n)) + MASK64 + 1n) & MASK64;
}

File: gdext/src/genome.cpp
/**
 * genome.cpp — Port de src/core/genome.ts
 *
 * Ver genome.h para la documentación completa y los invariantes de paridad.
 */

#include "genome.h"

#include <cassert>
#include <cstring>
#include <random>
#include <sstream>
#include <iomanip>

#ifdef _WIN32
#   include <windows.h>
#   include <bcrypt.h>
#   pragma comment(lib, "bcrypt.lib")
#else
#   include <fcntl.h>
#   include <unistd.h>
#endif

namespace petbits {

// ---------------------------------------------------------------------------
// Primitiva: extraer N bits a partir de offset
// ---------------------------------------------------------------------------

static inline uint64_t bits(uint64_t seed, int offset, int length) {
    const uint64_t mask = (length == 64) ? ~0ULL : ((1ULL << length) - 1ULL);
    return (seed >> offset) & mask;
}

// ---------------------------------------------------------------------------
// decodeGenome
// ---------------------------------------------------------------------------

Genes decodeGenome(Seed seed) {
    // Mapa de bits idéntico al GENOME_LAYOUT de genome.ts:
    //   lineage    offset=0  bits=4
    //   bodyShape  offset=4  bits=4
    //   eyes       offset=8  bits=4
    //   mouth      offset=12 bits=4
    //   appendages offset=16 bits=4
    //   pattern    offset=20 bits=4
    //   hue        offset=24 bits=8
    //   paletteMode  offset=32 bits=3
    //   temperament  offset=35 bits=3
    //   metabolism   offset=38 bits=3
    //   affinity     offset=41 bits=3
    //   proportion   offset=44 bits=4
    //   statBias     offset=48 bits=8
    //   mutation     offset=56 bits=8

    Genes g;
    g.lineage     = static_cast<uint8_t>(bits(seed,  0, 4));
    g.bodyShape   = static_cast<uint8_t>(bits(seed,  4, 4));
    g.eyes        = static_cast<uint8_t>(bits(seed,  8, 4));
    g.mouth       = static_cast<uint8_t>(bits(seed, 12, 4));
    g.appendages  = static_cast<uint8_t>(bits(seed, 16, 4));
    g.pattern     = static_cast<uint8_t>(bits(seed, 20, 4));
    g.hue         = static_cast<uint8_t>(bits(seed, 24, 8));
    g.paletteMode = static_cast<uint8_t>(bits(seed, 32, 3));
    g.temperament = static_cast<uint8_t>(bits(seed, 35, 3));
    g.metabolism  = static_cast<uint8_t>(bits(seed, 38, 3));
    g.affinity    = static_cast<uint8_t>(bits(seed, 41, 3));
    g.proportion  = static_cast<uint8_t>(bits(seed, 44, 4));

    const uint8_t statRaw = static_cast<uint8_t>(bits(seed, 48, 8));
    g.statBias.vigor   =  statRaw        & 0b11;
    g.statBias.animo   = (statRaw >> 2)  & 0b11;
    g.statBias.ingenio = (statRaw >> 4)  & 0b11;
    g.statBias.vinculo = (statRaw >> 6)  & 0b11;

    g.mutation = static_cast<uint8_t>(bits(seed, 56, 8));

    return g;
}

// ---------------------------------------------------------------------------
// formatSeed — "A3F0-91C4-77BE-2D08"
// ---------------------------------------------------------------------------

std::string formatSeed(Seed seed) {
    std::ostringstream oss;
    oss << std::uppercase << std::hex << std::setfill('0') << std::setw(16) << seed;
    const std::string hex = oss.str();
    return hex.substr(0, 4) + "-" + hex.substr(4, 4) + "-"
         + hex.substr(8, 4) + "-" + hex.substr(12, 4);
}

// ---------------------------------------------------------------------------
// hashString — FNV-1a 64 bits (mismo que hashString en genome.ts)
// ---------------------------------------------------------------------------

Seed hashString(std::string_view text) {
    uint64_t hash = 0xcbf29ce484222325ULL;
    constexpr uint64_t prime = 0x100000001b3ULL;
    for (unsigned char c : text) {
        hash ^= static_cast<uint64_t>(c);
        hash *= prime;
    }
    return hash;
}

// ---------------------------------------------------------------------------
// parseSeed — replica exacta de parseSeed() en genome.ts
// ---------------------------------------------------------------------------

Seed parseSeed(std::string_view input) {
    // Eliminar espacios y guiones
    std::string compact;
    for (char c : input) {
        if (c != ' ' && c != '-') compact += c;
    }

    if (compact.empty()) {
        throw std::invalid_argument("El seed no puede estar vacío");
    }

    // 0x... → hex explícito
    if (compact.size() > 2 && compact[0] == '0' && (compact[1] == 'x' || compact[1] == 'X')) {
        return std::stoull(compact, nullptr, 16);
    }

    // Todo dígito decimal
    bool allDigits = true;
    for (char c : compact) { if (!std::isdigit(c)) { allDigits = false; break; } }
    if (allDigits) {
        return std::stoull(compact, nullptr, 10);
    }

    // 1-16 caracteres hex (con letras a-f)
    bool allHex = (compact.size() >= 1 && compact.size() <= 16);
    bool hasLetter = false;
    for (char c : compact) {
        if (!std::isxdigit(c)) { allHex = false; break; }
        if (std::isalpha(c)) hasLetter = true;
    }
    if (allHex && hasLetter) {
        return std::stoull(compact, nullptr, 16);
    }
    if (allHex) {
        return std::stoull(compact, nullptr, 16);
    }

    // Texto arbitrario → hash FNV-1a (misma lógica que el TS)
    return hashString(input);  // con input original, no compact
}

// ---------------------------------------------------------------------------
// randomSeed — entropía criptográfica del sistema operativo
// ---------------------------------------------------------------------------

Seed randomSeed() {
    Seed seed = 0;
#ifdef _WIN32
    BCryptGenRandom(nullptr, reinterpret_cast<PUCHAR>(&seed), sizeof(seed),
                    BCRYPT_USE_SYSTEM_PREFERRED_RNG);
#else
    int fd = open("/dev/urandom", O_RDONLY);
    if (fd >= 0) {
        read(fd, &seed, sizeof(seed));
        close(fd);
    }
#endif
    return seed;
}

// ---------------------------------------------------------------------------
// Helpers de nombre
// ---------------------------------------------------------------------------

template<typename T>
static std::string_view nameFrom(const T& table, uint8_t index) {
    return table[index % table.size()];
}

std::string_view lineageName(const Genes& g)     { return nameFrom(LINEAGES,    g.lineage);    }
std::string_view temperamentName(const Genes& g) { return nameFrom(TEMPERAMENTS, g.temperament); }
std::string_view affinityName(const Genes& g)    { return nameFrom(AFFINITIES,  g.affinity);   }
std::string_view metabolismName(const Genes& g)  { return nameFrom(METABOLISMS, g.metabolism);  }

} // namespace petbits
